Feature/api endpoints

RileyManda opened the pull request:
The following milestone implementations has been completed according to stipulated [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_api_endpoints.md): Adding API Endpoints : 🌱 

# Implementations: 🧰  🔧 


### Api- End Points 🤝 

- Created an API endpoint to list all posts for a user.
- Created an API endpoint to list all comments for a user's post.
- Created an API endpoint to add a comment to a post.
- API endpoints receive JSON and respond JSON as well.


 In addition  following general requirements have been fulfilled:
 
### General requirements




> Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
> Made sure that I have used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
> Made sure that I have documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> Followed list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).


Please kindly review this implementation.

Thank you 🙏🏻

RileyManda committed:
Likes endpoints|create json endpoints for comments

RileyManda committed:
Bug fix: Comments and Posts

RileyManda committed:
Code cleanup

LionRouge1 commented on app/controllers/posts_controller.rb:
```diff
@@ -4,6 +4,10 @@ def index
     @posts = Post.all
     @user = User.find(params[:user_id])
     @posts = @user.posts
+    respond_to do |format|
+      format.html
+      format.json { render json: @posts }
+    end
   end
```
- [ ] The API endpoint to list all the posts created by a user isn't functioning as expected due to Devise authentication(see image). Kindly check and make sure the response is a JSON.
![Screenshot from 2023-10-12 19-25-15](https://github.com/RileyManda/blog-app/assets/91142432/5f0e2e0a-d425-44e0-87e6-94f0399e1396)

LionRouge1 commented on app/controllers/comments_controller.rb:
```diff
@@ -2,17 +2,27 @@ class CommentsController < ApplicationController
 
+  def index
+    @comments = @post.comments
+    respond_to do |format|
+      format.html
+      format.json { render json: @comments }
+    end
```
- [ ] The API endpoint to list all the posts' comments isn't functioning as expected due to Devise authentication(see image). Kindly check and make sure the response is a JSON.
![Screenshot from 2023-10-12 19-51-24](https://github.com/RileyManda/blog-app/assets/91142432/8d3e1a9c-14f9-4a4b-aed1-e7781a44a41f)

LionRouge1 commented on app/controllers/comments_controller.rb:
```diff
@@ -2,17 +2,27 @@ class CommentsController < ApplicationController
-      flash[:success] = 'Comment created successfully!'
-    else
-      flash[:error] = 'Error creating the comment.'
+    respond_to do |format|
+      if @comment.save
+        format.html { redirect_to user_post_path(@comment.post.author_id, @comment.post.id) }
+        format.json { render json: @comment, status: :created }
+      else
+        format.html { render :new }
+        format.json { render json: @comment.errors, status: :unprocessable_entity }
+      end
```
- [ ] Same for the API endpoint for adding a new comment to a post

LionRouge1 reviewed (changes requested):
Hi @RileyManda & @H-Ramalan, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

![almost](https://media.giphy.com/media/KyBesf8kNT6s88SEgb/giphy.gif)

## Highlights
- Good PR
- Good commit
- All linters passed

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

[OPTIONAL]
- [ ] It's a good practice to add `/api` to the URL or `api/v1` with the version number in order to identify it as an API-based URL and it also helps to separate the API's controllers from the other controllers. You can archive this by making use of the rails controller **Namespaces and routing**, you can read more about [namespace here](https://guides.rubyonrails.org/routing.html#controller-namespaces-and-routing)

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

H-Ramalan committed:
Api end-points:comments|posts

RileyManda committed:
Implemented serializers and api routes

RileyManda committed:
Bug fix:endpoints

RileyManda committed:
Linter fixes

RileyManda commented:
The following itertions have been implemented to the project: Api endpoints:

Iterations: 🔨 

- Fixed issue related to displaying comments and posts data

- added serializers

- Updated routes to include api path: api/v1/

<img width="1273" alt="Screen Shot 2023-10-13 at 11 40 48 AM" src="https://github.com/RileyManda/blog-app/assets/37291194/a6945c0b-5da5-42db-b268-d34563e22a60">
<img width="1271" alt="Screen Shot 2023-10-13 at 11 43 46 AM" src="https://github.com/RileyManda/blog-app/assets/37291194/d38f5e40-7353-4656-b6f9-50663c8ecdd9">


In addition the following general requirements have been fulfilled:
 
### General requirements




> Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
> Made sure that I have used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
> Made sure that I have documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> Followed list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).


Please kindly review this implementation.

Thank you 🙏🏻

addod19 reviewed (approved):
## Hi @RileyManda and @H-Ramalan, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

### Highlights
- [ ] Great job fixing the required changes as requested from the previous review :rocket: 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
